src/main/java/org/commoncrawl/spark/util/CCIndex2FilenameParser.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commoncrawl.spark.util;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CCIndex2FilenameParser {
+    protected static final Pattern filenameAnalyzer = Pattern
+                .compile("^(?:common-crawl/)?crawl-data/([^/]+)/segments/([^/]+)/(crawldiagnostics|robotstxt|warc|wat|wet)/");
+
+                // crawl-data/CC-NEWS/2019/01/CC-NEWS-20190101042830-00057.warc.gz
+    protected static final Pattern newsFilenameAnalyzer = Pattern
+                .compile("^(?:common-crawl/)?crawl-data/CC-NEWS/(\\d+)/(\\d+)/CC-NEWS-(.+)\\.warc\\.gz");
+
+    // Class to encapsulate the extracted crawl, segment, and subset.
+    public static class FilenameParts {
+        public String crawl;
+        public String segment;
+        public String subset;
+    }
+
+    // Error class if we can't find the crawl, segment, and subset.
+    public static class FilenameParseError extends Exception {
+        public FilenameParseError(String message) {
+            super(message);
+        }
+    }
+
+    public static FilenameParts getParts(String filename) throws FilenameParseError {
+        FilenameParts parts = new FilenameParts();
+        Matcher m = filenameAnalyzer.matcher(filename);
+        if(m.find()){
+            parts.crawl = m.group(1);
+            parts.segment = m.group(2);
+            parts.subset = m.group(3);
+        } else {
+            Matcher newsParts = newsFilenameAnalyzer.matcher(filename);
+            if(!newsParts.find()){
+                throw new FilenameParseError("Filename not parseable (tried normal and news): " + filename);
+            }
+            parts.crawl = String.format("CC-NEWS-%s-%s", newsParts.group(1), newsParts.group(2));
+            parts.segment = newsParts.group(3);
+            parts.subset = "news-warc";
```

Comment:
Why this shouldn't be just "warc"? The value in the "crawl" partition column already makes it clear that it's the news crawl and not the main crawl.

Reply:
Yes - I went back and forth on this a few times.
At the end of the day - it came down to a few comments Greg has made - in that we want to be extra careful about how users of the data may be sorting/filtering data, and not making changes that might upset their existing processes.

I'm not sure philosophically if we want to consider the news and main crawls "one thing" or if they should be separate.  So, I chose a middle ground, where news is in the same index/schema, but it wouldn't likely get filtered into scope by anyone looking for "warc" subset. (that's also why i chose to prefix with news rather than suffix - in case people might be doing "startswith warc" or "warc%" or something strange.

I realize people COULD filter news in/out based on the crawl value - but, again, I erred on the side of making the default for existing users "out".

Thoughts?

Reply:
> we want to be extra careful about how users of the data may be sorting/filtering data, and not making changes that might upset their existing processes.

I fully agree. There shouldn't be any changes which cannot be handled by schema merging. That is, extending the table is possible, but not changing the semantics of existing columns.

My assumption was that any new table (for CC-NEWS, the old crawls ARC files, but also the WAT/WET files) would break the semantics of the schema. That's why the location is:
``` 
s3://commoncrawl/cc-index/table/cc-main/warc/crawl=.../subset=.../
```
Consequently, a table for CC-NEWS would be on a different path. It would be a separate table which might use a different schema. Given the continuous release of the news crawl the path pattern could be:
```
s3://commoncrawl/cc-index/table/cc-news/warc/year=.../month=.../
```
The news crawl has a different collection method: it's sampled by time, not on a global level, with no revisits.

In addition, the cc-main table is already quite big. Simply for practical reasons, I would keep the news crawl and especially the WAT/WET records in a separate table. It's also more practical to announce separate new tables to the users than explain the extension of the existing one by new partitions and how to rephrase queries so that the results do not change.

Over the long term, an independent schema also allows for schema upgrades which focus on the given dataset format or scope: text metrics for WET or news-related metadata (publication time, etc.).